File: rake/fstar.rb
# $legal:598:
# 
# Copyright 2015 Michael Lowell Roberts
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# 
# ,$

require "pathname"

module Rake::FStar
   
   class << self
      include Rake::DSL

      def home_dir
         @home_dir ||= find_home_dir
      end

      def exe_path
         @exe_path ||= find_exe_path(self.home_dir)
      end

      def lib_path
         @lib_path ||= self.home_dir.join("lib")
      end

      def generate_tasks(files)
         task :verify => files do |t|
            sh "#{self.exe_path} --use_build_config #{t.prerequisites.join(" ")}"
         end
      end

      private

      def find_home_dir
         if ENV["FSTAR_HOME"].nil? then
            raise "i was unable to find the F* home directory; please set the FSTAR_HOME environment variable to the correct path (currently unassigned)."
         end
         home_dir = Pathname.new(ENV["FSTAR_HOME"])
         if not home_dir.exist? then
            raise "i was unable to find the F* home directory; please set the FSTAR_HOME environment variable to the correct path (currently \#{home_dir.to_s}\")"
         end
         home_dir
      end

      def find_exe_path(home_dir)
         pe = home_dir.join("bin/fstar.exe")
         elf = home_dir.join("bin/fstar")
         if pe.exist? then
            pe
         elsif elf.exist? then
            elf
         else
            raise "i was unable to find the F* executable (either \"#{pe.to_s}\" or \"#{elf.to_s}\"). is FSTAR_HOME (#{fstar_home.to_s}) set to the correct path?"
         end
      end

   end

end

# $vim-rb:31: vim:set sts=3 sw=3 et ft=ruby:,$

File: supplement/textadept/lexers/fstar.lua
-- $legal:629:
--
-- Copyright 2016 Michael Lowell Roberts & Microsoft Research
--
-- Licensed under the Apache License, Version 2.0 (the "License");
-- you may not use this file except in compliance with the License.
-- You may obtain a copy of the License at
--
--     http://www.apache.org/licenses/LICENSE-2.0
--
-- Unless required by applicable law or agreed to in writing, software
-- distributed under the License is distributed on an "AS IS" BASIS,
-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-- See the License for the specific language governing permissions and
-- limitations under the License.
--
--,$

-- adapted from the textadept F# lexer.

local l = require('lexer')
local token, word_match = l.token, l.word_match
local P, R, S = lpeg.P, lpeg.R, lpeg.S

local M = {_NAME = 'fstar'}

-- Whitespace.
local ws = token(l.WHITESPACE, l.space^1)

-- Comments.
local line_comment = P('//') * l.nonnewline^0
local block_comment = l.nested_pair('(*', '*)')
local comment = token(l.COMMENT, line_comment + block_comment)

-- Strings.
--local sq_str = l.delimited_range("'", true)
local dq_str = l.delimited_range('"', true)
--local string = token(l.STRING, sq_str + dq_str)
local string = token(l.STRING, dq_str)

-- Numbers.
local number = token(l.NUMBER, (l.float + l.integer * S('uUlL')^-1))

-- Preprocessor.
--[[
local preproc_word = word_match{
  'ifndef', 'ifdef', 'if', 'else', 'endif', 'light', 'region', 'endregion'
}
local preproc = token(l.PREPROCESSOR,
                      l.starts_line('#') * S('\t ')^0 * preproc_word *
                      (l.nonnewline_esc^1 + l.space * l.nonnewline_esc^0))
]]

-- Keywords.
local keyword = token(l.KEYWORD, word_match{
   'module', 'open',
   'type', 'and',
   'true', 'false',
   'Lemma', 'ensures', 'requires', 'decreases',
   'forall', 'exists',
   'val', 'let', 'rec', 'in',
   'abstract', 'inline', 'irreducible', 'private', 'unfoldable',
   'fun', 'function',
   'match', 'with',
   'if', 'then', 'else',
   'true', 'false', 'or', 'not',
   'assert', 'admit'
})

-- Types.
local type = token(l.TYPE, word_match{
  'Type',
  'True', 'False',
  'Tot', 'ML',
  'bool', 'int', 'nat', 'unit'
})

-- Identifiers.
local identifier = token(l.IDENTIFIER, l.word)

-- Operators.
local operator = token(l.OPERATOR, S('=<>+-*/\\^.,:;~!@#%^&|?[](){}'))

M._rules = {
  {'whitespace', ws},
  {'keyword', keyword},
  {'type', type},
  {'identifier', identifier},
  {'string', string},
  {'comment', comment},
  {'number', number},
  {'operator', operator},
}

return M
